pkg/etcdcli/etcdcli.go

```diff
@@ -0,0 +1,182 @@
+package etcdcli
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/operatorclient"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/etcdserver/etcdserverpb"
+	"go.etcd.io/etcd/pkg/transport"
+	"google.golang.org/grpc"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+type etcdClientGetter struct {
+	nodeLister      corev1listers.NodeLister
+	endpointsLister corev1listers.EndpointsLister
+
+	nodeListerSynced      cache.InformerSynced
+	endpointsListerSynced cache.InformerSynced
+}
+
+func NewEtcdClientGetter(kubeInformers v1helpers.KubeInformersForNamespaces) EtcdClient {
+	return &etcdClientGetter{
+		nodeLister:            kubeInformers.InformersFor("").Core().V1().Nodes().Lister(),
+		endpointsLister:       kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Endpoints().Lister(),
+		nodeListerSynced:      kubeInformers.InformersFor("").Core().V1().Nodes().Informer().HasSynced,
+		endpointsListerSynced: kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Endpoints().Informer().HasSynced,
+	}
+}
+
+func (g *etcdClientGetter) getEtcdClient() (*clientv3.Client, error) {
+	if !g.nodeListerSynced() {
+		return nil, fmt.Errorf("node lister not synced")
+	}
+	if !g.endpointsListerSynced() {
+		return nil, fmt.Errorf("node lister not synced")
+	}
+
+	etcdEndpoints := []string{}
+	nodes, err := g.nodeLister.List(labels.Set{"node-role.kubernetes.io/master": ""}.AsSelector())
```

Comment:
So we blindly add any master node as an endpoint? Why not just what etcd.svs reports as healthy? Depending on what we do down the road with the client we should try to dial in the backend.

What about if nothing reports healthy from svs endpoints we fall back on nodes and see if we have any luck? Just an observation.